finality-aleph/src/network/tests.rs

```diff
@@ -0,0 +1,425 @@
+use super::*;
+use crate::KEY_TYPE;
+use futures::{
+    channel::{mpsc, oneshot},
+    SinkExt,
+};
+use sc_network::{Event, ObservedRole, PeerId, ReputationChange};
+use sp_keystore::{testing::KeyStore, CryptoStore};
+use sp_runtime::traits::Block as BlockT;
+use std::{collections::HashSet, time::Duration};
+use substrate_test_runtime::Block;
+
+type Channel<T> = (
+    Arc<Mutex<mpsc::UnboundedSender<T>>>,
+    Arc<Mutex<mpsc::UnboundedReceiver<T>>>,
+);
+
+fn channel<T>() -> Channel<T> {
+    let (tx, rx) = mpsc::unbounded();
+    (Arc::new(Mutex::new(tx)), Arc::new(Mutex::new(rx)))
+}
+
+#[derive(Clone)]
+struct TestNetwork<B: BlockT> {
+    event_sinks: Arc<Mutex<Vec<mpsc::UnboundedSender<Event>>>>,
+    oneshot_sender: Arc<Mutex<Option<oneshot::Sender<()>>>>,
+    report_peer: Channel<(PeerId, ReputationChange)>,
+    disconnect_peer: Channel<(PeerId, Cow<'static, str>)>,
+    send_message: Channel<(PeerId, Cow<'static, str>, Vec<u8>)>,
+    announce: Channel<(B::Hash, Option<Vec<u8>>)>,
+    add_set_reserved: Channel<(PeerId, Cow<'static, str>)>,
+    remove_set_reserved: Channel<(PeerId, Cow<'static, str>)>,
+}
+
+impl<B: BlockT> TestNetwork<B> {
+    fn new(tx: oneshot::Sender<()>) -> Self {
+        TestNetwork {
+            event_sinks: Arc::new(Mutex::new(vec![])),
+            oneshot_sender: Arc::new(Mutex::new(Some(tx))),
+            report_peer: channel(),
+            disconnect_peer: channel(),
+            send_message: channel(),
+            announce: channel(),
+            add_set_reserved: channel(),
+            remove_set_reserved: channel(),
+        }
+    }
+}
+/*
+impl<B: BlockT> Clone for TestNetwork<B> {
+    fn clone(&self) -> Self {
+        TestNetwork {
+            event_sinks: self.event_sinks.clone(),
+            oneshot_sender: self.oneshot_sender.clone(),
+            report_peer: self.report_peer.clone(),
+            disconnect_peer: self.disconnect_peer.clone(),
+            send_message: self.send_message.clone(),
+            announce: self.announce.clone(),
+            add_set_reserved: self.add_set_reserved.clone(),
+            remove_set_reserved: self.remove_set_reserved.clone(),
+        }
+    }
+}*/
```

Comment:
Can get rid of that commented out piece :)

Reply:
Done